src/picongpu/include/fields/background/templates/bgrTWTS.hpp

```diff
@@ -0,0 +1,953 @@
+            unit_length(UNIT_LENGTH), auto_tdelay(auto_tdelay)
+        {
+            /* Note: These objects cannot be instantiated on CUDA GPU device. Since this is done
+                     on host (see fieldBackground.param), this is no problem. */
+            const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
+            halfSimSize=subGrid.getGlobalDomain().size / 2;
+        }
+        
+        HDINLINE PMacc::math::Vector<float3_X, FieldE::numComponents>
+        TWTSFieldE::dummyCast(const PMacc::math::Vector<float3_X, FieldE::numComponents>& x) const
+        { return x; }
+
+        HDINLINE PMacc::math::Vector<float3_X, FieldE::numComponents>
+        TWTSFieldE::dummyCast(const PMacc::math::Vector<float2_X, FieldE::numComponents>& x) const
+            {
+                PMacc::math::Vector<float3_X, FieldE::numComponents> result;
+
+                for( uint32_t i = 0; i < numComponents; ++i ) // cellIdx Ex, Ey and Ez
+                {
+                    result[i]=float3_X( (x[i]).x(),(x[i]).y(),float_X(0.0) );
+                }
+                return result;
+            }
+        
+        template<>
+        HDINLINE PMacc::math::Vector<float3_64,FieldE::numComponents>
+        TWTSFieldE::getEfieldPositions_SI<DIM3>(const DataSpace<simDim>& cellIdx) const
+        {
+            /* Note: Neither direct precisionCast on picongpu::cellSize
+                     or casting on floatD_ does work. */
+            floatD_X cellDim;
+            for (uint32_t i = 0; i<simDim;++i) cellDim[i]=picongpu::cellSize[i];
+                //const floatD_X cellDim = picongpu::cellSize;
+                const float3_64 cellDimensions = precisionCast<float_64>( dummyCast(cellDim) )
+                                                 * unit_length;
+                
+                /* TWTS laser coordinate origin is centered transversally and defined longitudinally
+                 * by the laser center (usually maximum of intensity) in y. */
+                float3_X laserOrigin = dummyCast( floatD_X(halfSimSize) );
+                laserOrigin[1]=float_X( focus_y_SI/cellDimensions.y() );
+                
+                /* For the Yee-Cell shifted fields, obtain the fractional cell index components and
+                 * add that to the total cell indices. The physical field coordinate origin is
+                 * transversally centered with respect to the global simulation volume. */
+                PMacc::math::Vector<float3_X, FieldE::numComponents> eFieldPositions = 
+                                    dummyCast(fieldSolver::NumericalCellType::getEFieldPosition());
+                
+                PMacc::math::Vector<float3_64,FieldE::numComponents> eFieldPositions_SI;
+                
+                for( uint32_t i = 0; i < FieldE::numComponents; ++i ) // cellIdx Ex, Ey and Ez
+                {
+                    eFieldPositions[i]   += ( dummyCast( floatD_X(cellIdx) ) - laserOrigin );
```

Comment:
I know and it's not wrong but sub-ideal.
positions should always be of type `floatD_` but I know that continues in other parts of the implementation.